### PR TITLE
kafka actions

### PR DESCRIPTION
**Pull request by h1alexbel**
@l3r8yJ take a look please

---

## PR-Codex overview
This PR adds Kafka actions and their related tests to the project.

### Detailed summary
- Added `Action` interface.
- Added `Wakeup` implementation of `Action`.
- Added `CommitSync` implementation of `Action`.
- Added `CommitAsync` implementation of `Action`.
- Added unit tests for `Wakeup`, `CommitSync` and `CommitAsync`.

> The following files were skipped due to too many changes: `src/test/java/io/github/eocqrs/kafka/act/CommitSyncTest.java`, `src/test/java/io/github/eocqrs/kafka/act/CommitAsyncTest.java`, `src/test/java/io/github/eocqrs/kafka/act/AssignPartitionsTest.java`, `src/main/java/io/github/eocqrs/kafka/act/AssignPartitions.java`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by zoeself** [comment]
@h1alexbel thank you for your Pull Request. I'll assign someone to review it soon. 

 If this PR solves a ``todo`` from the code, please don't forget to remove it.

**Comment by zoeself** [comment]
@h1alexbel please review this Pull Request. Deadline (when it should be merged or closed) is ``2023-06-05T14:22:18.218787``.

You should check if the requirements have been implemented (partially or in full), if there are unit tests covering the changes and if the CI build passes. Feel free to reject the PR or ask for changes if it's too big or not clear enough.

Estimation here is ``30 minutes``, that's how much you will be paid. You will be paid even if this PR gets rejected.

**Comment by zoeself** [comment]
@h1alexbel Don't forget to close this ticket before the deadline (2023-06-05T14:22:18). You are past the first half of the allowed period.

**Comment by zoeself** [comment]
@l3r8yJ Don't forget to close this ticket before the deadline (2023-06-08T14:29:09). You are past the first half of the allowed period.

File: src/main/java/io/github/eocqrs/kafka/act/ShutdownHook.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  @Override
+  public void apply() {
+    Runtime.getRuntime().addShutdownHook(
+      new Thread(() -> {
+        this.action.apply();
+        try {
+          this.main.join();
+        } catch (final InterruptedException ex) {
+          ex.printStackTrace();
```
Comment (l3r8yJ) [inline src/main/java/io/github/eocqrs/kafka/act/ShutdownHook.java:57]: exception swallowing, bad practice
```suggestion
          throw new RuntimeException(ex);
```

File: src/test/java/io/github/eocqrs/kafka/act/CommitSyncTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  /*
+   * @todo #325:45/DEV Introduce Integration Tests for Sync Commit
+   */
+  @Test
+  void commitsSync(
+    @Mock final KafkaConsumer<String, String> mck
+  ) {
+    final Action sync = new CommitSync(mck);
+    Assertions.assertDoesNotThrow(sync::apply);
```
Comment (l3r8yJ) [inline src/test/java/io/github/eocqrs/kafka/act/CommitSyncTest.java:52]: I think such tests are too weak, we can create a puzzle to create more difficult tests

File: src/main/java/io/github/eocqrs/kafka/act/Action.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+package io.github.eocqrs.kafka.act;
+
+/**
+ * Kafka Action.
+ *
+ * @author Aliaksei Bialiauski (abialiauski.dev@gmail.com)
+ * @since 0.2.5
+ */
+public interface Action {
```
Comment (l3r8yJ) [inline src/main/java/io/github/eocqrs/kafka/act/Action.java:31]:
```suggestion
@FunctionalInterface
public interface Action {
```

**Review by l3r8yJ — CHANGES_REQUESTED** [review]
@h1alexbel take a look at my comments below, please

File: src/test/java/io/github/eocqrs/kafka/act/CommitSyncTest.java (hunk 2 above)
Comment (h1alexbel) [reply src/test/java/io/github/eocqrs/kafka/act/CommitSyncTest.java:52]: @l3r8yJ totally agree, I created them for integration tests

File: src/test/java/io/github/eocqrs/kafka/act/ShutdownHookTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+   */
+  @Test
+  void shutdownsWithWakeUp(
+    @Mock final KafkaConsumer<String, String> mck
+  ) {
+    final Action hook =
+      new ShutdownHook(
+        new Wakeup(mck),
+        Thread.currentThread(),
+        new Action() {
```
[the remark this answers is not in the document]
Comment (h1alexbel) [reply src/test/java/io/github/eocqrs/kafka/act/ShutdownHookTest.java]: applied

**Review by l3r8yJ — APPROVED** [review]
@h1alexbel lgtm, but Rultor does not want to merge because of the coverage, need to bump

**Comment by h1alexbel** [comment]
@l3r8yJ now, we need test for `ShutdownHook`'s catch section

**Comment by zoeself** [comment]
@h1alexbel I couldn't find any assignee for this task. This is either because there are no contributors with role ``REV`` available or because the project does not have enough funds.

Please, make sure there is at least one available contributor with the required role and the project can afford to pay them.

**Comment by h1alexbel** [comment]
@l3r8yJ I remove the shutdown hook for now, see #356

ready to review again

**Review by l3r8yJ — APPROVED** [review]
@h1alexbel lgtm!

**Comment by l3r8yJ** [comment]
@h1alexbel maybe, after merging this, a version of `2.4.1' will be released?

**Comment by h1alexbel** [comment]
@l3r8yJ I planned to release `0.2.5`

**Comment by h1alexbel** [comment]
@rultor merge

**Pull request merged**